matrix/basic_matrix.py

```diff
@@ -189,32 +292,46 @@ def is_hankel(self):
 
         return True
 
+    def is_symplectic(self):
+        """ Checks whether matrix is symplectic.  Returns True if symplectic matrix, False otherwise """
+
+        if not self.is_square():
+            return False
+
+        n = self.len_col // 2
+
+        ls_identity = self.identity_ls_entries(n)
+        ls_identity_neg = [[-1 * i for i in inner] for inner in ls_identity]
+        ls_zeros = self.zero_ls_entries(n, n)
+
+        # block matrix must be non-singular, skew-symmetric
+        # Use the typical choice:
+        # [[0  I_n],
+        # [-I_n 0]]
+        block_matrix = Matrix(self.create_block_ls_entries(top_left=Matrix(ls_zeros), top_right=Matrix(ls_identity),
+                                                           bottom_left=Matrix(ls_identity_neg),
+                                                           bottom_right=Matrix(ls_zeros)))
+
+        transpose = Matrix(ls_entries=[[self[j][i] for j in range(self.len_row)] for i in range(self.len_col)])
+
+        if transpose._mul(block_matrix)._mul(self) == block_matrix:
+            return True
+
+        return False
+
     def is_orthogonal(self):
         """
         Test whether column is orthogonal
         """
 
-        # Duplicates code for the identity matrix and then to multiply to avoid max recursion issues in matrix factory
-        # due to the identity matrix being orthogonal
-
         # Get identity matrix
-        ls_zero = self.zero_ls_entries(row_dim=self.len_col, col_dim=self.len_col)
-        for i in range(self.len_col):
-            ls_zero[i][i] = 1
-        I = Matrix(ls_zero)
+        ls_identity = self.identity_ls_entries(self.len_col)
+        I = Matrix(ls_identity)
 
         transpose = Matrix(ls_entries=[[self[j][i] for j in range(self.len_row)] for i in range(self.len_col)])
 
         # Multiply transpose with self and check whether it is equal to the identity matrix
-        result = self.zero_ls_entries(row_dim=self.len_row, col_dim=transpose.len_col)
-        for i in range(self.len_row):
-            # iterate through columns of Y
-            for j in range(transpose.len_col):
-                # iterate through rows of Y
-                for k in range(transpose.len_row):
-                    result[i][j] += self[i][k] * transpose[k][j]
-
-        return Matrix(result) == I
+        return transpose._mul(self) == I
```

Comment:
I'm wondering if we need to worry about floating point precision at all or if the matrix equality methods already take care of that for us.

Reply:
Will make this change! :)